Replace cluster linearization algorithm with SFL

sipa opened the pull request:
Part of cluster mempool: #30289.

This replaces the cluster linearization algorithm introduced in #30126 and #30286 (a combination of LIMO with candidate-set search), with a completely different algorithm: [spanning-forest linearization](https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419/1), which appears to have much better performance for hard clusters. See [this post](https://delvingbitcoin.org/t/how-to-linearize-your-cluster/303/68) for a comparison between various linearization algorithms, and [this post](https://delvingbitcoin.org/t/how-to-linearize-your-cluster/303/73) for benchmarks comparing them. Replaying historical mempool data on it shows that it can effectively linearize every observed cluster up to 64 transactions optimally within tens of microseconds, though pathological examples can be created which take longer.

The algorithm is effectively a very specialized version of the [simplex algorithm](https://en.wikipedia.org/wiki/Simplex_algorithm) to the problem of finding high-feerate topological subsets of clusters, but modified to find all consecutive such subsets concurrently rather than just the first one. See the post above for how it is related.

It represents the cluster as partitioned into a set of chunks, each with a spanning tree of its internal dependencies connecting the transactions. Randomized improvements are made by selecting dependencies to add and remove to these spanning trees, merging and splitting chunks, until no more improvements are possible, or a computation budget is reached. Like simplex, it does not necessarily make progress in every step, and thus has no upper bound on its runtime to find optimal, but randomization makes long runtimes very unlikely, and additionally makes it hard to adversarially construct clusters in which the algorithm reliably makes bad choices.

sipa committed:
clusterlin: add known-correct optimal linearization tests (tests)

sipa committed:
clusterlin: replace benchmarks with SFL-hard ones (bench)

This also adds a per-cost variant of each.

jonatack commented on src/cluster_linearize.h:
```diff
@@ -539,492 +555,651 @@ class LinearizationChunking
 };
 
-/** Class encapsulating the state needed to find the best remaining ancestor set.
+/** Class to represent the internal state of the spanning-forest linearization algorithm.
```
Appreciate the excellent doxygen documentation here.

jonatack reviewed (commented):
Concept ACK

ismaelsadeeq commented:
> I'm interested in seeing benchmarks of this on various systems, especially high-end/modern ones (as their performance is likely most predictive of what future hardware will be like).

Below are benchmarks on MacBook M2 PRO 2023.
Compiled using clang with only build bench cmake option.
```
cmake -B build -DBUILD_BENCH=ON && cmake --build build
```

<details>
<summary>master 9a887baadebce7c2e76df831fad54c5fa81d309e</summary>

```terminal
abubakarismail@Abubakars-MacBook-Pro ~/D/W/b/bitcoin (master)> ./build/bin/bench_bitcoin --filter="Linearize.*Example.*" -min-time=1
0000
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|           59,940.13 |           16,683.31 |    1.6% |     12.42 | `LinearizeOptimallyExample00`
|           87,998.62 |           11,363.81 |    1.3% |     10.91 | `LinearizeOptimallyExample01`
|           47,459.31 |           21,070.68 |    3.5% |      7.95 | `LinearizeOptimallyExample02`
|           50,438.60 |           19,826.08 |    0.7% |     10.96 | `LinearizeOptimallyExample03`
|            2,399.35 |          416,780.31 |    0.7% |      9.86 | `LinearizeOptimallyExample04`
|           13,914.66 |           71,866.62 |    0.7% |     11.13 | `LinearizeOptimallyExample05`
|           34,236.58 |           29,208.52 |    2.4% |      9.72 | `LinearizeOptimallyExample06`
|            9,453.32 |          105,782.94 |    0.5% |     12.12 | `LinearizeOptimallyExample07`
|           59,616.91 |           16,773.76 |    3.0% |     10.28 | `LinearizeOptimallyExample08`
|           12,952.64 |           77,204.32 |    1.2% |     11.71 | `LinearizeOptimallyExample09`
|            9,479.22 |          105,493.97 |    1.2% |     10.80 | `LinearizeOptimallyExample10`
|      269,154,375.00 |                3.72 |    1.0% |     11.32 | `LinearizeOptimallyExample11`
|          495,909.35 |            2,016.50 |    1.2% |     11.09 | `LinearizeOptimallyExample12`
|           52,703.02 |           18,974.25 |    4.7% |     10.48 | `LinearizeOptimallyExample13`
|          339,836.74 |            2,942.59 |    2.9% |     12.26 | `LinearizeOptimallyExample14`
|       12,825,954.92 |               77.97 |    1.8% |     10.89 | `LinearizeOptimallyExample15`
|        7,987,753.88 |              125.19 |    1.1% |     11.02 | `LinearizeOptimallyExample16`
|        9,785,196.49 |              102.20 |    0.6% |     11.02 | `LinearizeOptimallyExample17`
|       10,269,746.38 |               97.37 |    0.6% |     11.25 | `LinearizeOptimallyExample18`
|       11,814,387.50 |               84.64 |    1.1% |     11.10 | `LinearizeOptimallyExample19`
</details>



<details>
<summary> origin/pr/32545 55931c3cddee570d1740713760493dce654b6d34</summary>

```terminal
abubakarismail@Abubakars-MacBook-Pro ~/D/W/b/bitcoin ((55931c3c))> ./build/bin/bench_bitcoin --filter="Linearize.*Example.*" -min-ti
me=1000
```

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,957.10 |          510,960.04 |    1.3% |      1.10 | `LinearizeOptimallyExample00`
|            3,379.21 |          295,927.52 |    1.3% |      1.09 | `LinearizeOptimallyExample01`
|            9,121.03 |          109,636.77 |   38.7% |      0.89 | :wavy_dash: `LinearizeOptimallyExample02` (Unstable with ~7,962.5 iters. Increase `minEpochIterations` to e.g. 79625)
|            7,037.58 |          142,094.25 |    2.6% |      1.05 | `LinearizeOptimallyExample03`
|           15,914.11 |           62,837.34 |    1.1% |      1.10 | `LinearizeOptimallyExample04`
|           13,188.59 |           75,823.12 |    0.6% |      1.10 | `LinearizeOptimallyExample05`
|           13,121.48 |           76,210.91 |    1.7% |      1.07 | `LinearizeOptimallyExample06`
|           10,790.78 |           92,671.74 |    1.0% |      1.09 | `LinearizeOptimallyExample07`
|           13,731.89 |           72,823.19 |    0.7% |      1.10 | `LinearizeOptimallyExample08`
|           19,934.82 |           50,163.47 |    0.5% |      1.10 | `LinearizeOptimallyExample09`
|           12,701.04 |           78,733.73 |    0.7% |      1.10 | `LinearizeOptimallyExample10`
|           18,066.61 |           55,350.74 |    0.2% |      1.10 | `LinearizeOptimallyExample11`
|           25,834.49 |           38,707.95 |    0.2% |      1.10 | `LinearizeOptimallyExample12`
|           29,825.58 |           33,528.26 |    0.8% |      1.10 | `LinearizeOptimallyExample13`
|           38,191.18 |           26,184.06 |    0.4% |      1.10 | `LinearizeOptimallyExample14`
|           38,278.41 |           26,124.39 |    0.8% |      1.10 | `LinearizeOptimallyExample15`
|           45,135.55 |           22,155.49 |    0.8% |      1.11 | `LinearizeOptimallyExample16`
|           62,035.19 |           16,119.88 |    0.5% |      1.06 | `LinearizeOptimallyExample17`
|           64,734.11 |           15,447.81 |    0.4% |      1.06 | `LinearizeOptimallyExample18`
|           78,911.71 |           12,672.39 |    0.6% |      1.08 | `LinearizeOptimallyExample19`
</details>

Almost 10x speedup 🚀

fjahr commented:
> I'm interested in seeing benchmarks of this on various systems

<details>
<summary>Macbook Pro M4</summary>

$ build/bin/bench_bitcoin --filter="Linearize.*Example.*" -min-time=10000

|             ns/cost |              cost/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                1.12 |      893,382,447.85 |    0.5% |     11.06 | `LinearizeBoundedExample0`
|                1.34 |      743,601,759.92 |    0.7% |     11.07 | `LinearizeBoundedExample1`
|                1.14 |      880,692,144.98 |    0.3% |     10.99 | `LinearizeBoundedExample2`
|                1.33 |      749,663,261.00 |    0.3% |     10.98 | `LinearizeBoundedExample3`
|                1.24 |      807,545,779.18 |    0.5% |     10.57 | `LinearizeBoundedExample4`
|                1.37 |      730,046,090.82 |    0.4% |     10.64 | `LinearizeBoundedExample5`
|                1.28 |      779,862,831.19 |    0.4% |     10.66 | `LinearizeBoundedExample6`
|                1.17 |      856,937,081.55 |    0.4% |     10.71 | `LinearizeBoundedExample7`
|                1.19 |      840,624,085.43 |    0.2% |     10.74 | `LinearizeBoundedExample8`
|                1.19 |      840,938,970.51 |    0.4% |     10.74 | `LinearizeBoundedExample9`

|               ns/op |                op/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|            1,155.37 |          865,524.35 |    0.2% |     11.00 | `LinearizeOptimallyExample00`
|            2,170.00 |          460,828.59 |    0.2% |     11.04 | `LinearizeOptimallyExample01`
|            4,507.21 |          221,866.90 |    0.2% |     11.01 | `LinearizeOptimallyExample02`
|            4,435.85 |          225,435.87 |    0.3% |     10.96 | `LinearizeOptimallyExample03`
|           10,189.93 |           98,136.12 |    0.2% |     11.02 | `LinearizeOptimallyExample04`
|            8,854.57 |          112,936.09 |    0.4% |     10.79 | `LinearizeOptimallyExample05`
|            8,685.08 |          115,140.05 |    0.1% |     10.82 | `LinearizeOptimallyExample06`
|            7,162.17 |          139,622.57 |    0.0% |     10.64 | `LinearizeOptimallyExample07`
|            9,120.88 |          109,638.49 |    0.0% |     11.00 | `LinearizeOptimallyExample08`
|           12,780.73 |           78,242.79 |    0.1% |     11.00 | `LinearizeOptimallyExample09`
|            9,242.70 |          108,193.53 |    0.1% |     10.98 | `LinearizeOptimallyExample10`
|           13,653.81 |           73,239.62 |    0.3% |     11.04 | `LinearizeOptimallyExample11`
|           19,211.80 |           52,051.34 |    0.2% |     11.00 | `LinearizeOptimallyExample12`
|           22,243.75 |           44,956.45 |    0.1% |     11.00 | `LinearizeOptimallyExample13`
|           28,409.81 |           35,199.11 |    0.2% |     10.99 | `LinearizeOptimallyExample14`
|           28,265.23 |           35,379.16 |    0.1% |     10.99 | `LinearizeOptimallyExample15`
|           33,767.58 |           29,614.20 |    0.3% |     11.02 | `LinearizeOptimallyExample16`
|           46,851.80 |           21,343.90 |    0.4% |     11.03 | `LinearizeOptimallyExample17`
|           48,470.35 |           20,631.17 |    0.1% |     10.98 | `LinearizeOptimallyExample18`
|           57,668.81 |           17,340.40 |    0.4% |     10.98 | `LinearizeOptimallyExample19`

</details>

l0rinc commented:
I've run the command on a `Raspberry Pi 4 Model B - Cortex-A72 4-Core ARM with SanDisk SSD PLUS 1TB (USB 3.0)` with to have data about lower-end hardware as well. 

<details>
<summary>full command</summary>

```bash
git remote add sipa https://github.com/sipa/bitcoin.git || true && git fetch sipa bench_sfl_css && git switch -C bench_sfl_css sipa/bench_sfl_css && \
git clean -fxd && git reset --hard && \
cmake -B build -DBUILD_BENCH=ON -DCMAKE_BUILD_TYPE=Release && cmake --build build -j"$(nproc)" && \
wget -q https://bitcoin.sipa.be/clusters/clusters.tgz && tar -xzf clusters.tgz && \
for F in clusters_sim2023_large clusters_medium clusters_spanning clusters_bipartite; do \
    time build/bin/bench_bitcoin --filter="LinearizeDataSet" <"$F" >"$F.out"; \
done && \
tar -czf clusters.out.tgz clusters_*.out
```

</details>

It took considerably longer than expected (~13h):
* `clusters_sim2023_large`: 11h 39m
* `clusters_medium`: 26m 33s
* `clusters_spanning`: 19m 37s
* `clusters_bipartite`: 37m 17s

Let's see if I can attach it here: [clusters.out.tgz](https://github.com/user-attachments/files/20589411/clusters.out.tgz)

l0rinc commented:
> Also, is this with 32-bit or 64-bit userspace?

64-bit userspace (AArch64)
```bash
$ file bitcoind
bitcoind: ELF 64-bit LSB pie executable, ARM aarch64, version 1 (GNU/Linux), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, BuildID[sha1]=7e059ec01f7460042910ca4ed15270382269c9d5, for GNU/Linux 3.7.0, with debug_info, not stripped
```

<details>
<summary>additional details</summary>

```bash
$ getconf LONG_BIT
64
$ dpkg --print-architecture
arm64
$ uname -m
aarch64
```

</details>

sipa committed:
clusterlin: remove unused MergeLinearizations (cleanup)

This ended up never being used in txgraph.

sipa committed:
clusterlin: drop support for improvable chunking (simplification)

With MergeLinearizations() gone and the LIMO-based Linearize() replaced by SFL, we do not
need a class (LinearizationChunking) that can maintain an incrementally-improving chunk
set anymore.

Replace it with a function (ChunkLinearizationInfo) that just computes the chunks as
SetInfos once, and returns them as a vector. This simplifies several call sites too.

sipa committed:
clusterlin: randomize equal-feerate parts of linearization (privacy)

This places equal-feerate chunks (with no dependencies between them) in random
order in the linearization output, hiding information about DepGraph insertion
order from the output. Likewise, it randomizes the order of transactions within
chunks for the same reason.

sipa committed:
clusterlin: keep FIFO queue of improvable chunks (preparation)

This introduces a queue of chunks that still need processing, in both
MakeTopological() and OptimizationStep(). This is simultaneously:
* A preparation for introducing randomization, by allowing permuting the
  queue.
* An improvement to the fairness of suboptimal solutions, by distributing
  the work more fairly over chunks.
* An optimization, by avoiding retrying chunks over and over again which
  are already known to be optimal.

sipa commented:
This PR has grown quite a bit in scope (despite still being a net negative in LoC!), so I don't think it's unreasonable to split it up into an SFL-specific one, and one with follow-up changes to `txgraph`, or even further. I'll wait for reviewer comments.

sipa committed:
clusterlin: replace cluster linearization with SFL (feature)

This replaces the existing LIMO linearization algorithm (which internally uses
ancestor set finding and candidate set finding) with the much more performant
spanning-forest linearization algorithm.

This removes the old candidate-set search algorithm, and several of its tests,
benchmarks, and needed utility code.

The worst case time per cost is similar to the previous algorithm, so
ACCEPTABLE_ITERS is unchanged.

marcofleon reviewed (commented):
Did a first pass of the changes in `cluster_linearize.h`, left a couple small comments. I'm running a few of the fuzz targets, including the new one, and I'll leave those going for a while.

Still need to look at the tests thoroughly. Let me know if there's anything specific you think reviewers can do that would be useful.

instagibbs commented:
glancing through the PR now

how bad would it be to merge all the way through "clusterlin: replace cluster linearization with SFL implementation (feature)", maybe some of the commits that drops unused things (ala "clusterlin: remove unused MergeLinearizations (cleanup)") and defer the other optimisations for next PR?

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -653,6 +669,686 @@ class AncestorCandidateFinder
+ * The sets of transactions that are internally connected by active dependencies are called chunks.
+ * Each chunk of N transactions contains exactly N-1 active dependencies (an additional one would
+ * necessarily form a cycle), and thus those active dependencies form a spanning tree for the chunk.
+ * The collection of all spanning trees for the entire cluster form a spanning forest. Each
```
```Suggestion
 * The collection of all spanning trees for the entire cluster form a spanning forest. In the extreme each
```

sipa commented on src/cluster_linearize.h:
```diff
@@ -386,6 +386,22 @@ struct SetInfo
+    }
+
+    /** Compute the difference between this and other SetInfo (which must be a subset). */
+    SetInfo operator-(const SetInfo& other) noexcept
```
Done.

sipa commented:
Now a proper repository for the Python version: https://github.com/sipa/pyclusterlin

sipa commented on src/cluster_linearize.h:
```diff
@@ -653,6 +669,686 @@ class AncestorCandidateFinder
+                    // Skip if this dependency is ineligible (the top chunk that would be created
+                    // does not have higher feerate than the chunk it is currently part of).
+                    if (!(dep_data.top_setinfo.feerate >> chunk_data.chunk_setinfo.feerate)) continue;
+                    // Activate it otherwise.
```
Done.

sipa commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -1130,6 +1134,103 @@ FUZZ_TARGET(clusterlin_simple_linearize)
+        auto lin = sfl.GetLinearization();
+        // Which must be valid.
+        SanityCheck(depgraph, lin);
+        // If we started from a topological input, the resulting feerate diagram cannot be worse.
```
Done.

sipa commented on src/cluster_linearize.h:
```diff
@@ -653,6 +669,686 @@ class AncestorCandidateFinder
+ *
+ * Each chunk has a feerate: the total fee of all transactions in it divided by the total size of
+ * all transactions in it. We say the spanning forest is topological whenever no inactive
+ * dependencies exist from one chunk to another chunk with lower or equal feerate. The algorithm
```
Do you mean the write-up in the code comment added here, on the one on Delving?

Do you think it's worth elaborating on here?

sipa commented:
I have pushed an additional WIP commit which rewords and expands the big `SpanningForestState` comment. LMK what you think. If reviewers think it's an improvement, I'll squash it into the individual commits.

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -461,96 +461,123 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, std::
+ *
+ *                The algorithm attempts to keep the state topological as much as possible, so it
+ *                can be interrupted to produce an output whenever, but will sometimes need to
+ *                temporarily deviate from it when improving the state.
```
> temporarily deviate from it when improving the state.

I'm expecting each `Improve()` call to result in a topological state, assuming one exists, and that this means during a call

instagibbs reviewed (commented):
8c990cc2f5a7e93a5debb7eaf20e7f2c887df70d seems like an improvement, only a couple comments

marcofleon reviewed (commented):
I think this new comment is better.

sipa commented on src/cluster_linearize.h:
```diff
@@ -461,96 +461,123 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, std::
+ *                two distinct chunks such that the child chunk has higher or equal feerate than
+ *                the parent chunk.
+ *
+ *                The relevance is that whenever the state is topological, the produced output
```
It's not, but it's pretty subtle.

You could have two equal-feerate chunks that depend on each other in one way but not the other one. If so, it doesn't matter whether they're merged or not.

Worth elaborating on?

instagibbs replied:
ok, so I just "forgot" again that "or equal" means you can clearly have same feerate chunks that can be merged, hence not topological in the SFL state sense

sipa commented on src/cluster_linearize.h:
```diff
@@ -461,96 +461,123 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, std::
- * linearization is also topological.
+ * The algorithm consists of switching dependencies between active and inactive. The final
+ * linearization that is produced at the end consists of these chunks, sorted from high to low
+ * feerate, each individually sorted in an arbitrary but topological (= no parent before child)
```
Thanks, fixed!

sipa commented on src/cluster_linearize.h:
```diff
@@ -461,96 +461,123 @@ std::vector<FeeFrac> ChunkLinearization(const DepGraph<SetType>& depgraph, std::
+ *            graph, ignoring the parent/child direction. This is equivalent to saying that within
+ *            each chunk the set of active dependencies form a tree, and thus the overall set of
+ *            active dependencies in the graph form a spanning forest, giving the algorithm its
+ *            name. Being acylic is also equivalent to every chunk of N transactions having exactly
```
Fixed.

sipa committed:
clusterlin: ReadLinearization for non-topological (tests)

Rather than using an ad-hoc no-dependency copy of the graph when a potentially
non-topological linearization is needed in the clusterlin fuzz test, add this
directly as a feature in ReadLinearization().

This is preparation for a later commit where another use for such a function
is added.

sipa committed:
clusterlin: add class implementing SFL state (preparation)

This adds a data structure representing the optimization state for the spanning-forest
linearization algorithm (SFL), plus a fuzz test for its correctness.

This is preparation for switching over Linearize() to use this algorithm.

See https://delvingbitcoin.org/t/spanning-forest-cluster-linearization/1419 for
a description of the algorithm.

sipa committed:
clusterlin: add support for loading existing linearization (feature)

sipa committed:
clusterlin: randomize various decisions in SFL (feature)

This introduces a local RNG inside the SFL state, which is used to randomize
various decisions inside the algorithm, in order to make it hard to create
pathological clusters which predictably have bad performance.

The decisions being randomized are:
* When deciding what chunk to attempt to split, the queue order is
  randomized.
* When deciding which dependency to split on, a uniformly random one is
  chosen among those with higher top feerate than bottom feerate within
  the chosen chunk.
* When deciding which chunks to merge, a uniformly random one among those
  with the higher feerate difference is picked.
* When merging two chunks, a uniformly random dependency between them is
  now activated.
* When making the state topological, the queue of chunks to process is
  randomized.

instagibbs commented on src/bench/cluster_linearize.cpp:
```diff
@@ -210,40 +210,48 @@ void BenchMergeLinearizationsWorstCase(DepGraphIndex ntx, benchmark::Bench& benc
+            total_cost += cost;
+        }
+
+        // Benchmark the time per cost.
```
094e845f7309c8cf4e172c8fe07e200e054ca2db

TIL, nice

instagibbs commented on src/cluster_linearize.h:
```diff
@@ -662,6 +678,697 @@ class AncestorCandidateFinder
+    };
+
+    /** The set of all transactions in the cluster. */
+    SetType m_transactions;
```
c2eaf22503f7af057599f02ecb5f24a267ea33e2

suggestion:
```
-    /** The set of all transactions in the cluster. */
-    SetType m_transactions;
-    /** Information about each transaction (and chunks). Indexed by TxIdx. */
+    /** The set of all TxIdx's of transactions in the cluster indexing into m_tx_data. */
+    SetType m_transaction_idxs;
+    /** Information about each transaction (and chunks). Keeps the "holes" from DepGraph
+        during construction. Indexed by TxIdx. */
     std::vector<TxData> m_tx_data;
```

sipa replied:
Done, except renaming to `m_transaction_idxs`, because I'm too lazy to update all the affected commits right now. Will leave this open.

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -328,18 +328,22 @@ SetType ReadTopologicalSet(const DepGraph<SetType>& depgraph, const SetType& tod
+                }
             }
+        } else {
+            potential_next = todo;
```
9f7f80f9407d10c95c2690e34dfa918c6a361f21

nit

```Suggestion
            // Allow any element to be selected next, regardless of topology
            potential_next = todo;
```

instagibbs reviewed (commented):
reviewed through ab1416bb471c9326bb77ee464be6aeebf663e024

continuing on with optimizations next

only minor quibbles after the nice reworkings of comments and UpdateChunk

sipa commented on src/test/cluster_linearize_tests.cpp:
```diff
@@ -52,6 +54,63 @@ void TestDepGraphSerialization(const std::vector<std::pair<FeeFrac, SetType>>& c
+                FixLinearization(depgraph, lin);
+                break;
+            }
+            std::tie(lin, opt, cost) = Linearize(depgraph, 1000000, rng.rand64(), lin);
```
I have just increased it to 1T iterations, which ought to be enough for anyone.

sipa commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -1230,6 +1054,13 @@ FUZZ_TARGET(clusterlin_sfl)
+    // Verify that optimality is reached within an expected amount of work.
+    //
+    if (sfl.GetCost() > MaxOptimalLinearizationIters(depgraph.TxCount())) {
+        assert(is_optimal);
```
The point here is to prevent against potential bugs added in the future, which result in optimality just never being reached anymore. Absent bounds on expected iteration count, this would render a large portion of this test powerless (because anything in `is_optimal`, which includes all quality checks) won't be reached anymore.

Added a comment. Does this explain it?

sipa commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -1192,13 +1196,11 @@ FUZZ_TARGET(clusterlin_sfl)
         auto lin = sfl.GetLinearization();
         // Which must be valid.
         SanityCheck(depgraph, lin);
-        // If we started from a topological input, the resulting feerate diagram must be at least
```
This was a result of commit surgery gone wrong. Fixed.

brunoerg reviewed (commented):
I ran an incremental mutation test for this PR, specifically for `src/cluster_linearize.h` (which generated 222 mutants), considering unit (`cluster_linearize_tests`), functional (`mempool_packages`), and fuzzing.

For fuzzing, existing targets were run with qa-assets inputs. For the `clusterlin_sfl` target, I created a simple corpus from a few hours of executions (not very reliable).

Overall, the tests are excellent, achieving a mutation rate of over 98% (excluding equivalent mutants and some useless ones). I've commented out some unkilled mutants here; feel free to ignore them or add tests to eliminate them if it makes sense.

Happy to run it again if any relevant changes are made.

instagibbs commented on src/txgraph.cpp:
```diff
@@ -1311,13 +1311,13 @@ void SingletonClusterImpl::AppendChunkFeerates(std::vector<FeeFrac>& ret) const
     // Iterate over the chunks of this cluster's linearization.
-    for (unsigned i = 0; i < linchunking.NumChunksLeft(); ++i) {
-        const auto& [chunk, chunk_feerate] = linchunking.GetChunk(i);
+    for (unsigned i = 0; i < linchunking.size(); ++i) {
```
26661a6c655ad148d3ecf200c64b89c439b6d44a

nit: while we're here (we also arent using `j` below anywhere, but meh)

```Suggestion
    for (const auto& [chunk, chunk_feerate] : linchunking) {
```

instagibbs reviewed (commented):
reviewed through 26661a6c655ad148d3ecf200c64b89c439b6d44a

sorry all I have is ticky-tacky, will go one more pass over before finishing

instagibbs commented on src/test/fuzz/cluster_linearize.cpp:
```diff
@@ -1169,6 +890,104 @@ FUZZ_TARGET(clusterlin_simple_linearize)
+    try {
+        reader >> rng_seed >> flags >> Using<DepGraphFormatter>(depgraph);
+    } catch (const std::ios_base::failure&) {}
+    if (depgraph.TxCount() <= 1) return;
```
Feel like we should still cover size 1, even if it's trivial. Thoughts?

instagibbs reviewed (approved):
ACK edc64813b221598994da52cc5b7124a02ab7e042

Much more straightforward than the in-master linearization approach, even ignoring its clear performance benefits.

marcofleon reviewed (commented):
ACK edc64813b221598994da52cc5b7124a02ab7e042

This implementation was easier for me to review/understand than the current one on master, which I spent some time with when reviewing https://github.com/bitcoin/bitcoin/pull/30605. We'll see if I still feel that way after looking at the optimizations.

I ran the `clusterlin_sfl` and `clusterlin_linearize` fuzz targets for a while and no issues came up. Here's the coverage: [sfl](https://marcofleon.github.io/coverage/clusterlin_sfl/) and [linearize](https://marcofleon.github.io/coverage/clusterlin_linearize/).

I also verified that the [test vectors](https://github.com/sipa/pyclusterlin/blob/main/linearization_tests.json) in the python repo are the same ones in the unit test. Both tests pass and arrive at optimal linearizations for the chosen mempool clusters.

Left a couple non-blocking comments that can be addressed in the follow up, if desired.

sipa commented:
Oops, sorry, I had to revert this last change because the property I claimed (each new transaction traversed in `LoadLinearization()` is in its own chunk still) only holds if the input linearization is topological. With that, I don't see an effectively way of asserting that `LoadLinearization()` is called in the right state, without introducing a performance penalty.

```diff
diff --git a/src/cluster_linearize.h b/src/cluster_linearize.h
index 7b74d433722..d65e61d4458 100644
--- a/src/cluster_linearize.h
+++ b/src/cluster_linearize.h
@@ -395,7 +395,7 @@ struct SetInfo
         return *this;
     }
 
-    /** Remove the transactions of other from this SetInfo (must be subset). */
+    /** Remove the transactions of other from this SetInfo (which must be a subset). */
     SetInfo& operator-=(const SetInfo& other) noexcept
     {
         Assume(other.transactions.IsSubsetOf(transactions));
@@ -951,11 +951,7 @@ public:
     {
         // Add transactions one by one, in order of existing linearization.
         for (DepGraphIndex tx : old_linearization) {
-            // Since this function must be called after construction, each new transaction added
-            // is in its own singleton chunk still.
-            auto chunk_rep = tx;
-            // Verify that it is indeed in its own chunk.
-            Assume(m_tx_data[tx].chunk_rep == chunk_rep);
+            auto chunk_rep = m_tx_data[tx].chunk_rep;
             // Merge the chunk upwards, as long as merging succeeds.
             while (true) {
                 chunk_rep = MergeStep<false>(chunk_rep);
```
